src/Razor/Microsoft.AspNetCore.Razor.Tools/src/RewriteCssCommand.cs

```diff
@@ -146,19 +141,39 @@ protected override void VisitSelector(Selector selector)
                 //   ".first child," containing two simple selectors: ".first" and "child"
                 //   ".second", containing one simple selector: ".second"
                 // Our goal is to insert immediately after the final simple selector within each selector
-                var lastSimpleSelector = selector.Children.OfType<SimpleSelector>().LastOrDefault();
+
+                // If there's a deep combinator among the sequence of simple selectors, we consider that to signal
+                // the end of the set of simple selectors for us to look at, plus we strip it out
+                var allSimpleSelectors = selector.Children.OfType<SimpleSelector>();
+                var firstDeepCombinator = allSimpleSelectors.FirstOrDefault(s => IsDeepCombinator(s.Text));
+
+                var lastSimpleSelector = allSimpleSelectors.TakeWhile(s => s != firstDeepCombinator).LastOrDefault();
                 if (lastSimpleSelector != null)
                 {
-                    InsertionPositions.Add((lastSimpleSelector.AfterEnd, ScopeInsertionType.Selector));
+                    Edits.Add(new InsertSelectorScopeEdit { Position = lastSimpleSelector.AfterEnd });
+                }
+
+                // Also remove the deep combinator if we matched one
+                if (firstDeepCombinator != null)
+                {
+                    Edits.Add(new DeleteContentEdit { Position = firstDeepCombinator.Start, DeleteLength = firstDeepCombinator.Length });
                 }
             }
 
+            private static bool IsDeepCombinator(string simpleSelectorText)
+            {
+                // ::deep is a convenient and friendly syntax
+                // ::razor-deep is an alias the people can use if they are concerned about possible clashes with other meanings
+                return string.Equals(simpleSelectorText, "::deep", StringComparison.Ordinal)
```

Comment:
👍

Reply:
Note, I removed `::razor-deep` because I realised it doesn't make any sense. These tokens are already stripped out by the time the CSS reaches the browser, so it's not possible to clash against anything in the browser. The place you *could* clash is if you wanted the token *not* to be interpreted as a deep combinator by our rewriter. But adding more aliases doesn't give you any way to do that, so it was pointless.

In the future we could expose some option for what the combinator text is, but there's no requirement for that today.